### PR TITLE
Load resources for 404 page if the page doesn't exist

### LINKED ISSUE #7252: 404 page flashes and then the page becomes blank for pages that don't exist
## Description

(I am using Gatsby v2)

When trying to load a page that doesn't exist, the 404 page flashes and then the page becomes blank.

#2223 is similar, but I think this is a different problem because `render` isn't being called for my `Layout` component.

I did some poking around and I believe that this happens because `getResourcesForPathname` is always called with `location.pathname`, so when `PageRenderer` is rendered none of the resources for the pages are loaded and it render's null. (I could be reading things wrong though)

### Steps to reproduce

Go to: https://blissful-bassi-4d7d8c.netlify.com/something and observe the blank page

That website is just a copy of [gatsby-starter-default v2](https://github.com/gatsbyjs/gatsby-starter-default/tree/v2)

### Expected result

The 404 page should be displayed

### Actual result

A blank page is ultimately displayed

### Environment

I set up a repro on netlify with the latest `@next` packages to reproduce:

```
    "gatsby": "^2.0.0-beta.98",
    "gatsby-plugin-manifest": "^2.0.2-beta.5",
    "gatsby-plugin-offline": "^2.0.0-beta.9",
    "gatsby-plugin-react-helmet": "^3.0.0-beta.4",
```

https://github.com/haroldhues/test-gatsby-starter-default-404/tree/master

(FYI `gatsby info --clipboard` failed on my computer (Windows 10)  with an unhanded rejection...)

### File contents (if changed)

`gatsby-config.js`: N/A 
`package.json`: N/A 
`gatsby-node.js`: N/A 
`gatsby-browser.js`: N/A 
`gatsby-ssr.js`: N/A

### PR DESCRIPTION
**Pull request by CodyRay**
Fixes #7252

Also I changed `gatsby serve` to serve `404.html` if pages were missing, similar to the way many hosting platforms do.

I couldn't find existing unit test coverage for these files, Is there a place where I should add unit/automated tests?

### CODE REVIEW

**Comment by KyleAMathews** [comment]
Deploy preview for *using-postcss-sass* failed.

Built with commit ecedd55c9f233bb201e599e9e56525cb6f718661

https://app.netlify.com/sites/using-postcss-sass/deploys/5b70c11082d3f1419d9b9d45

File: docs/docs/how-to-contribute.md [hunk 1]
```diff
@@ -58,7 +58,7 @@ The usual contributing steps are:
 - Setup up repo and install dependencies: `yarn run bootstrap`
 - Make sure tests are passing for you: `yarn test`
 - Create a topic branch: `git checkout -b topics/new-feature-name`
-- Run `npm run watch` from the root of the repo to watch for changes to packages' source code and compile these changes on-the-fly as you work. Note that the watch command can be resource intensive. To limit it to the packages you're working on, add a scope flag, like `npm run watch -- --scope={gatsby,gatsby-cli}`. To watch just one package, run `npm run watch -- --scope=gatsby`.
+- Run `yarn run watch` from the root of the repo to watch for changes to packages' source code and compile these changes on-the-fly as you work. Note that the watch command can be resource intensive. To limit it to the packages you're working on, add a scope flag, like `yarn run watch -- --scope={gatsby,gatsby-cli}`. To watch just one package, run `yarn run watch -- --scope=gatsby`.
```
Comment (LeKoArts) [inline docs/docs/how-to-contribute.md]: Why did you change the npm commands?

Comment (haroldhues) [reply docs/docs/how-to-contribute.md]: It seemed inconsistent, everywhere else in this doc `yarn` was used instead of `npm`

I just assumed it was overlooked, happy to drop that commit as well

Comment (LeKoArts) [reply docs/docs/how-to-contribute.md]: > ensure you have the latest version of yarn installed (>= 1.0.2)

Ahh, I guess it's fine then 👍

Comment (m-allanson) [reply docs/docs/how-to-contribute.md]: The docs _do_ use npm everywhere, except in these contributing docs where people are asked to use yarn instead.
